girder/utility/plugin_utilities.py

```diff
@@ -17,462 +14,94 @@
-    allPlugins = findAllPlugins()
-    dag = {}
-    visited = set()
-
-    if plugins is None:
-        plugins = set(allPlugins.keys())
-    else:
-        plugins = set(plugins)
-
-    def addDeps(plugin):
-        if plugin not in allPlugins:
-            message = 'Required plugin %s does not exist.' % plugin
-            if ignoreMissing:
-                logprint.error(message)
-                return
-            else:
-                raise ValidationException(message)
-
-        deps = set()
-        for key in keys:
-            deps |= allPlugins[plugin][key]
-        dag[plugin] = deps
-
-        for dep in deps:
-            if dep in visited:
-                return
-            visited.add(dep)
-            addDeps(dep)
+from stevedore.extension import ExtensionManager
+from stevedore.named import NamedExtensionManager
+from toposort import toposort_flatten
 
-    for plugin in plugins:
-        addDeps(plugin)
+NAMESPACE = 'girder.plugin'
 
-    for pset in toposort(dag):
-        for plugin in pset:
-            yield plugin
 
-
-def loadPlugin(name, root, appconf, apiRoot=None):
+class config(object):  # noqa: class name
     """
-    Loads a plugin into the application. This means allowing it to create
-    endpoints within its own web API namespace, and to register its event
-    listeners, and do anything else it might want to do.
+    Wrap a plugin's ``load`` method appending plugin metadata.
 
-    :param name: The name of the plugin (i.e. its directory name)
-    :type name: str
-    :param root: The root node of the web API.
-    :param appconf: The cherrypy configuration for the server.
-    :type appconf: dict
+    :param name str: The plugin's name
+    :param description str: A brief description of the plugin.
+    :param version str: A semver compatible version string.
+    :param dependencies list: A list of plugins required by this plugin.
+    :returns: A decorator that appends the metadata to the method
     """
-    if apiRoot is None:
-        apiRoot = root.api.v1
-
-    pluginDir = os.path.join(getPluginDir(), name)
-    isPluginDir = os.path.isdir(os.path.join(pluginDir, 'server'))
-    isPluginFile = os.path.isfile(os.path.join(pluginDir, 'server.py'))
-    pluginLoadMethod = None
+    def __init__(self, **kw):
+        self.config = kw
 
-    if not os.path.exists(pluginDir):
-        # Try to load the plugin as an entry_point
-        for entry_point in iter_entry_points(group='girder.plugin', name=name):
-            pluginLoadMethod = entry_point.load()
-            module = importlib.import_module(entry_point.module_name)
-            pluginDir = os.path.dirname(module.__file__)
-            module.PLUGIN_ROOT_DIR = pluginDir
-            girder.plugins.__dict__[name] = module
-            isPluginDir = True
+    def __call__(self, func):
+        @six.wraps(func)
+        def wrapped(*arg, **kw):
+            return func(*arg, **kw)
+        wrapped.config = self.config
+        return wrapped
```

Comment:
This config decorator was defined in https://github.com/girder/girder/pull/1255.  It is used to decorate the entry_point function and is intended to replace the `plugin.json` files.  I have left the original implementation for now.